### PR TITLE
Remove dynamic dispatch from (General)LazyBufferCache

### PR DESCRIPTION
**Pull request by LilithHafner**
## Checklist

- [ ] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API

### CODE REVIEW

File: src/PreallocationTools.jl [hunk 1]
```diff
@@ -204,18 +204,17 @@ same type and size `f(size(u))` (defaulting to the same size), which is allocate
...
+    bufs::Dict{Any, Any} # a dictionary mapping (type, size) pairs to buffers
     sizemap::F
     LazyBufferCache(f::F = identity) where {F <: Function} = new{F}(Dict(), f) # start with empty dict
 end
 
 # override the [] method
 function Base.getindex(b::LazyBufferCache, u::T) where {T <: AbstractArray}
     s = b.sizemap(size(u)) # required buffer size
-    buf = get!(b.bufs, (T, s)) do
+    get!(b.bufs, (T, s)) do
```
Comment (ChrisRackauckas) [inline src/PreallocationTools.jl:215]: Does that change actually make a difference?

Comment (LilithHafner) [reply src/PreallocationTools.jl:215]: Not at all. That is just for stylistic consistency with the rest of the file.

File: docs/src/index.md [hunk 2]
```diff
@@ -208,12 +208,9 @@ A `LazyBufferCache` is a `Dict`-like type for the caches, which automatically de
 new cache arrays on demand when they are required. The function `f` maps
 `size_of_cache = f(size(u))`, which by default creates cache arrays of the same size.
 
-Note that `LazyBufferCache` does cause a dynamic dispatch, though it is type-stable.
-This gives it a ~100ns overhead, and thus on very small problems it can reduce
-performance, but for any sufficiently sized calculation (e.g. >20 ODEs) this
-may not be even measurable. The upside of `LazyBufferCache` is that the user does
-not have to worry about potential issues with chunk sizes and such: `LazyBufferCache`
-is much easier!
+Note that `LazyBufferCache` is type-stable and contains no dynamic dispatch. This gives
```
Comment (ChrisRackauckas) [inline docs/src/index.md:211]: doesn't it have an internal dynamic dispatch?

Comment (LilithHafner) [reply docs/src/index.md:211]: No, it's just a lookup from a `Dict{Any, Any}`, which is not type stable, but doesn't have dynamic dispatch either.

Verified via `@btime` indicating no allocations and via `@profview` being a happy color (dynamic dispatch is supposed to show up in red)

Pasted image 3

(In theory the correct way to verify is JET)

Comment (ChrisRackauckas) [reply docs/src/index.md:211]: wow that's pretty incredible. Julia used to not be able to omit that.

**Pull request merged**